File: 1538-maximum-points-you-can-obtain-from-cards/maximum-points-you-can-obtain-from-cards.js
/**
 * @param {number[]} cardPoints
 * @param {number} k
 * @return {number}
 */
var maxScore = function(cardPoints, k) {
    if(k === cardPoints.length) {
        let sum = 0;
        return cardPoints.reduce((acc, curr) => acc + curr, sum);
    }
    let leftSum = 0;
    let rightSum = 0; 
    let maxPoints = 0;
    for(let i = 0; i < k; i++) {
        leftSum += cardPoints[i];
    }
    maxPoints = leftSum;
    let rightIndex = cardPoints.length - 1;
    for(let i = k-1; i >= 0; i--) {
        leftSum -= cardPoints[i];
        rightSum += cardPoints[rightIndex];
        rightIndex -= 1;
        maxPoints = Math.max(maxPoints, leftSum + rightSum);
    }
    return maxPoints;
};

File: 1538-maximum-points-you-can-obtain-from-cards/maximum-points-you-can-obtain-from-cards.java
class Solution {
    public int maxScore(int[] cardPoints, int k) { // [1,2,3,4,5,6,1], 4
        int n = cardPoints.length; // 7
        int leftSum = 0; // 0
        int l = 0; // 0
        int r = 0; // 0
        while(r < k) { // 4 < 4
            leftSum += cardPoints[r]; // 10
            r += 1; // 4
        }
        if(n == k) return leftSum; // 7 == 4; NO
        int rightSum = 0; // 0
        int max = leftSum; // 10
        l = r-1; // 
        r = n-1;
        while( (n - r) + (l + 1) != k && l >= 0 && r >= 0) {
            leftSum = leftSum - cardPoints[l];
            l -= 1;
            rightSum = rightSum + cardPoints[r];
            max = Math.max(max, leftSum+rightSum);
            r -= 1;
        }
        return max;
    }
}